src/envoy/http/authn/authenticator_base.cc

```diff
@@ -71,17 +70,48 @@ void AuthenticatorBase::validateX509(
     done_callback(&payload, false);
   }
 
-  // TODO (lei-tang): Adding other attributes (i.e ip) to payload if desire.
+  // TODO (lei-tang): Adding other attributes (i.e ip) to payload if needed.
   done_callback(&payload, true);
 }
 
+// TODO (lei-tang): Extracting more attributes/claims to payload if needed.
+void extractJwtPayload(std::map<std::string, std::string>& jwt_payload,
+                       Payload* payload) {
+  // Extract attributes from JWT payload
+  // Extract user
+  if (jwt_payload.count("iss") > 0 && jwt_payload.count("sub") > 0) {
+    payload->mutable_jwt()->set_user(jwt_payload["iss"] + "/" +
+                                     jwt_payload["sub"]);
+  }
+  if (jwt_payload.count("aud") > 0) {
+    payload->mutable_jwt()->add_audiences(jwt_payload["aud"]);
+  }
+  // Extract authorized presenter (azp)
+  if (jwt_payload.count("azp") > 0) {
+    payload->mutable_jwt()->set_presenter(jwt_payload["azp"]);
+  }
+}
+
 void AuthenticatorBase::validateJwt(
     const iaapi::Jwt&,
-    const AuthenticatorBase::MethodDoneCallback& done_callback) const {
+    const AuthenticatorBase::MethodDoneCallback& done_callback) {
   Payload payload;
-  // TODO (diemtvu/lei-tang): construct jwt_authenticator and call Verify;
-  // pass done_callback so that it would be trigger by jwt_authenticator.onDone.
-  done_callback(&payload, false);
+  Envoy::Http::HeaderMap& header = *filter_context()->headers();
+  const Network::Connection* connection = filter_context()->connection();
+  std::map<std::string, std::string> jwt_payload;
```

Comment:
JwtPayload proto has a map<string, string> for (raw) claims. You can use it instead of this separate map. The raw claims need to be copied to payload message anyway.

Reply:
Done.